pocld/daemon.cc

```diff
@@ -415,6 +421,137 @@ int PoclDaemon::launch(std::string ListenAddress, struct ServerPorts &Ports) {
   return 0;
 }
 
+#ifdef ENABLE_VSOCK
+static int convertToVsockAddress(const std::string &VsockContextId, int Port,
+                                 struct sockaddr *Sockaddr,
+                                 socklen_t *SockaddrLen) {
+  if (Sockaddr == nullptr || SockaddrLen == nullptr)
+    return -1;
+
+  char *End = nullptr;
+  int Sibling = 0;
+  long CidValue;
+  struct sockaddr_vm *svm = (struct sockaddr_vm *)Sockaddr;
+
+  if (VsockContextId.empty()) {
+    *SockaddrLen = sizeof(*svm);
+    memset(svm, 0, *SockaddrLen);
+    svm->svm_family = AF_VSOCK;
+    svm->svm_cid = VMADDR_CID_ANY;
+    svm->svm_port = Port;
+
+    return 0;
+  }
+
+  /*
+   * Connection to a sibling VM
+   * VsockContextId is of the form "s<cid>"
+   */
+  if (VsockContextId[0] == 's') {
+#if LINUX_VERSION_CODE > KERNEL_VERSION(5, 11, 0)
+    Sibling = 1;
+#endif
+  }
+
+  CidValue = Sibling ? strtol(VsockContextId.c_str() + 1, &End, 10)
+                     : strtol(VsockContextId.c_str(), &End, 10);
+  if (VsockContextId.c_str() != End && *End == '\0' && CidValue <= UINT_MAX) {
+    *SockaddrLen = sizeof(*svm);
+    memset(svm, 0, *SockaddrLen);
+    svm->svm_family = AF_VSOCK;
+    svm->svm_cid = CidValue;
+    svm->svm_port = Port;
+#if LINUX_VERSION_CODE > KERNEL_VERSION(5, 11, 0)
+    if (Sibling) {
+      svm->svm_flags = VMADDR_FLAG_TO_HOST;
+    }
+#endif
+
+    return 0;
+  }
+  return -1;
+}
+
+int PoclDaemon::launchVsock(std::string ListenAddress,
+                            struct ServerPorts &Ports) {
+  ListenPorts = {Ports};
+  LastSessionId = 0;
+  pid_t ServerPid = getpid();
+  int Error = 0;
+  NumListenFds = 0;
+  int One = 1;
+
+  struct sockaddr_vm ServerAddrCommand, ServerAddrStream;
+  socklen_t VsockAddrLen;
+  struct sockaddr VsockAddr;
+
+  Error = convertToVsockAddress(ListenAddress, ListenPorts.command, &VsockAddr,
```

Comment:
Same here, would it be possible / sensible to merge this with `pocl_resolve_address` so we can reuse the code path that just takes a `struct sockaddr*` without caring what kind of address it actually is?

Reply:
getaddrinfo canot support AF_VSOCK, it seems more sensible to merge this with pocl_resolve_address, I will try fill getaddrinfo manually